library/std/src/io/copy.rs

```diff
@@ -56,12 +56,47 @@ where
 
 /// The general read-write-loop implementation of
 /// `io::copy` that is used when specializations are not available or not applicable.
-pub(crate) fn generic_copy<R: ?Sized, W: ?Sized>(reader: &mut R, writer: &mut W) -> io::Result<u64>
+pub(crate) fn generic_copy<R: ?Sized, W: ?Sized>(reader: &mut R, writer: &mut W) -> Result<u64>
 where
     R: Read,
     W: Write,
 {
-    let mut buf = MaybeUninit::<[u8; super::DEFAULT_BUF_SIZE]>::uninit();
+    let writer_cap = WriteBuffer::capacity(writer);
+    if writer_cap > 0 {
+        // A non-zero capacity indicates that we have a BufWriter, even though we can't assert that through types
+        // We must flush the buffer because specialization in buffered_copy will bypass BufWriter, no matter which
+        // buffer we later pass as reference
+        WriteBuffer::prepare_buffer(writer)?;
+
+        if writer_cap >= super::DEFAULT_BUF_SIZE {
+            let raw_buf = WriteBuffer::raw_buffer(writer);
+            // SAFETY: raw_buf comes from a BufWriter that we just flushed, i.e. it is no longer
+            // referencing any of the bytes and will not be dereferenced for the duration of the copy operation
+            // because the WriteBuffer trait will be used to bypass it.
+            // The bytes may be uninitialized but this is handled by casting to `[MaybeUninit<u8>]` first.
+            let buf = unsafe { (raw_buf as *mut [MaybeUninit<u8>]).as_mut().unwrap() };
+
+            return buffered_copy(reader, writer, buf);
+        }
+    }
+
+    // Either no buffering wrappers are involved or the capacity is smaller than the default one
+    // Use a stack buffer instead
```

Comment:
```suggestion
    // Either no buffering wrappers are involved or the capacity is smaller than the default one.
    // Use a stack buffer instead.
```
We probably don't need period for one line comment but multi-line it would be good to place them I think it makes it clearer.